rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -85,6 +107,132 @@ class ReportingServiceImpl(
+
+  def nodeInfoService: NodeInfoService
+
+  /**
+   * The cache is managed node by node.
+   * A missing nodeId mean that the cache wasn't initialized for
+   * that node.
+   */
+  private[this] var cache = Map.empty[NodeId, Set[RuleNodeStatusReport]]
+
+  private[this] def cacheToLog(c: Map[NodeId, Set[RuleNodeStatusReport]]): String = {
+    //display compliance value and expiration date.
+    c.map { case (nodeId, reports) =>
+
+      val reportsString = reports.map { r =>
+        val run = r.agentRunTime match {
+          case None => "no agent run"
+          case Some(r) => s"last run@${r}"
+        }
+        val cfid = r.configId match {
+          case None => "no config id"
+          case Some(c) => s"configId:${c.value}"
+        }
+        s"${r.ruleId.value}/${r.serial}[exp:${r.expirationDate}][${run}][${cfid}]${r.compliance.toString}"
+      }.mkString("\n  ", "\n  ", "")
+
+      s"node: ${nodeId.value}${reportsString}"
+    }.mkString("\n", "\n", "")
+  }
+
+  /**
+   * For the nodeIds in parameter, check that the cache is:
+   * - initialized, else go find missing rule node status reports (one time for all)
+   * - none reports is expired, else switch its status to "missing" for all components
+   */
+  def checkAndUpdateCache(nodeIds: Set[NodeId]) : Box[Map[NodeId, Set[RuleNodeStatusReport]]] = this.synchronized {
+    if(nodeIds.isEmpty) {
+      Full(Map())
+    } else {
+      val now = DateTime.now
+
+      /*
+       * Three cases: 1/ cache does not exist, 2/ cache exists but expiration date expired, 3/ cache does note exists
+       * For simplicity (and keeping computation logic elsewhere than in a cache that already has its cache logic
+       * to manage), we will group 2 and 3, but it is well noted that it seems that a RuleNodeStatusReports that
+       * expired could be computed without any more logic than "every thing is missing".
+       *
+       */
+
+      val (expired, upToDate) = nodeIds.partition { id =>  //two group: expired id, and up to date info
+        cache.get(id) match {
+          case None =>
+            if(id == NodeId("36DB125C-8F69-11E4-B8D0-072EA8D53D78")) { println("not in cache ????????")}
+            true
+          case Some(set) =>
+            if( set.exists { _.expirationDate.isBefore(now) } ) {
+              if(id == NodeId("36DB125C-8F69-11E4-B8D0-072EA8D53D78")) { println("expired ????????")}
+              true
+            } else {
+              false
+            }
+        }
+      }
+      for {
+        newStatus <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(expired, Set())
+      } yield {
+        /*
+         * Here, it's missing all the UnexpectedVersion/NoInitNoVersion results, because, well, they
+         * actually don't have rulenodestatutreport. So.
+         * We add them in NodeId -> Set()
+         */
+        val updated = newStatus.groupBy { _.nodeId }
+        val invalidVersion = expired -- updated.keySet
```

Comment:
i don't understand how a node could not be in the result of findRuleNodeStatusReports; unless it has been deleted